doc/development/style_guide.md

```diff
@@ -0,0 +1,211 @@
+  For example, when creating a component that takes an `opacity` level as an input, you should check
+  whether the value is in the range from 0 to 1. Consider also including the value itself into the
+  error message, to make it easier for the developer to debug the error:
+  ```dart
+  assert(0 <= opacity && opacity <= 1, 'The opacity value must be from 0 to 1: $opacity');
+  ```
+  Always use asserts as early as possible to detect possible violations. For example, check the
+  validity of `opacity` in the constructor/setter, instead of in the render function.
+
+  When adding such an assert, also include a test that checks that the assert triggers. This test
+  would verify that the component does not accept invalid input, and that the error message is what
+  you expect it to be.
+
+- Use an assert without an error message to check for a condition that cannot be triggered by the
+  developer through any means known to you. If such an assert does trigger, it would indicate a bug
+  in the Flame framework.
+
+  Such asserts serve as "mini-tests" directly in the code, and protect against future refactorings
+  that could create an erroneous internal state. It should not be possible to write a test that
+  would deliberately trigger such an assert.
+
+- Use an explicit if-check with an exception to test for a condition that may be outside of the
+  developer's control (i.e. it may depend on the environment or on user's input). When deciding
+  whether to use an assert or exception, consider the following question: is it possible for the
+  error condition to occur in production even after the developer has done extensive testing on
+  their side?
+
+
+### Class structure
+
+- Consider putting all class constructors at the top of the class. This makes it much easier to see
+  how the class ought to be used.
+
+- Try to make as much of your class' API private as possible, do not expose members "just in case".
+  This makes it much easier to modify/refactor the class later without it being a breaking change.
+
+  Remember to document all your public members! Documenting things is harder than it looks, and one
+  way to avoid the burden of documentation is to make as many variables private as possible.
+
+- If a class exposes a `List<X>` or `Vector2` property -- it is **NOT** an invitation to modify
+  them at will! Consider such properties as read-only, unless the documentation explicitly says that
+  they are allowed to be modified.
+
+- When a class becomes sufficiently big, consider adding *regions* inside it, which help with code
+  navigation and collapsing (note the lack of space after `//`):
+  ```dart
+  //#region Region description
+  ...
+  //#endregion
+  ```
+
+- If a class has a private member that needs to be exposed via a getter/setter, prefer the following
+  code structure:
+  ```dart
+  class MyClass {
+    MyClass();
+
+    ...
+    int _variable;
```

Comment:
Don't we usually have this in the block below the dartdoc?

Reply:
So that all private variables would be in one block? It feels like it would be easier to find/see the variable if it was within the same block as the getters and setters, so either between them or under the setter.

Reply:
Come to think of it, I actually like having all private variables in one block. It gives you an ability to look at the class and at a glance see how light or heavy it is. As opposed to the current situation, where in order to see the size of the class, say `Component`, one needs to scroll through a thousand of lines of code.

Compared to that, finding the variable from its getter is both easy (a simple shortcut to go to variable's definition), and rarely needed (since the getter already tells you the name and the type of the variable).

Reply:
True, that is a good argument. Maybe add aline about all private variables being in one block then?